core/client/db/src/lib.rs

```diff
@@ -680,6 +691,10 @@ impl<Block> state_machine::ChangesTrieStorage<Blake2Hasher, NumberFor<Block>>
 where
 	Block: BlockT<Hash=H256>,
 {
+	fn cached_changed_keys(&self, root: &H256) -> Option<HashSet<Vec<u8>>> {
```

Comment:
Good point. But we need to clone items of this set (`Vec<u8>`) anyway. So this could only help with the container internals itself (that `HashSet`). Returning `RwLock` in public interfaces isn't a good practice, imo (simply because it reveals internals) + it is the `RwLock` for all blocks => we'll allow external users to hold the whole object, which is deadlock-prone. So probably passing closure to that fn will work. The closure will be called for every cached item. But I'm not sure - whether this'll lead to any "trait cannot be made into an object" error somewhere. I'll try && respond later here.

Reply:
I've replaced `cached_changed_keys() -> Option<HashSet>` to `with_cached_changed_keys(functor) -> bool`. Now `HashMap` isn't cloned

Reply:
It seems to me that even if we use a closure, the lock will get locked a bit longer than with cloning, so I am starting to wonder if it is fine.
I also realized that cache item are not modified at all, so using an Arc in front of HashSet seems like an ok solution to (I checked this here https://github.com/cheme/substrate/commit/8ddb0b562f80229456a70f45922ad9150c8d5631).

I also realized that cached item are actually only use once so 'remove' instead of 'get' can also be better (but it does not play as well: requires to switch some cache mutability).

Reply:
`Arc<HashMap<_, _>> is a bit confusing, imo :) The lock shouldn't last for too long + actually, at least in current implementation, we do not have concurrent block import => shouldn't be a problem at least for now.

Yeah - I had the same idea about `remove` instead of `get`. But it isn't perfect, because on authorities block is first built, and then - imported (afaik) - so if we'll remove when block is built, then on import cache will be empty => we'll need to read everything from db. Also - if block import fails, then we also may have the cleared cache. The latter, though, doesn't look like a big issue, though.